Consumer chain bootstrap

Currently bootstrapping ICS fails on the consumer with the following:

```sh
$ make consumer-start; make provider-start
```

The provider produces blocks just fine, but the consumer crashes at InitChain:

```go
4:23PM DBG running initialization for module module=auth
4:23PM DBG running initialization for module module=bank
4:23PM DBG running initialization for module module=distribution
4:23PM DBG running initialization for module module=staking
4:23PM DBG running initialization for module module=gov
4:23PM DBG running initialization for module module=mint
4:23PM DBG running initialization for module module=genutil
panic: failed to execute DeliverTx for '{
	          "body": {
	            "messages": [
	              {
	                "@type": "/cosmos.staking.v1beta1.MsgCreateValidator",
	                "description": {
	                  "moniker": "localnet",
	                  "identity": "",
	                  "website": "",
	                  "security_contact": "",
	                  "details": ""
	                },
	                "commission": {
	                  "rate": "0.100000000000000000",
	                  "max_rate": "0.200000000000000000",
	                  "max_change_rate": "0.010000000000000000"
	                },
	                "min_self_delegation": "1",
	                "delegator_address": "",
	                "validator_address": "cosmosvaloper1h300y74ufhh0fu8qg5feysvq282s270hy3a43s",
	                "pubkey": {
	                  "@type": "/cosmos.crypto.ed25519.PubKey",
	                  "key": "itpTY3nOygpriL5p6lXrqjLeLK7MAJ2HKB0wta0b2p0="
	                },
	                "value": {
	                  "denom": "uatone",
	                  "amount": "1000000000"
	                }
	              }
	            ],
	            "memo": "ee88f886a6691366429b17117142737eedcfed29@192.168.2.90:26656",
	            "timeout_height": "0",
	            "extension_options": [],
	            "non_critical_extension_options": []
	          },
	          "auth_info": {
	            "signer_infos": [
	              {
	                "public_key": {
	                  "@type": "/cosmos.crypto.secp256k1.PubKey",
	                  "key": "AvciriZUjjan66VVS1c242gHspFBbDkAMol2sxVu7K2O"
	                },
	                "mode_info": {
	                  "single": {
	                    "mode": "SIGN_MODE_DIRECT"
	                  }
	                },
	                "sequence": "0"
	              }
	            ],
	            "fee": {
	              "amount": [],
	              "gas_limit": "200000",
	              "payer": "",
	              "granter": ""
	            },
	            "tip": null
	          },
	          "signatures": [
	            "rRgUelpsHQ4qi1y48uDtvXoocZqPlPITQ4ef6Jl25WIWa9eIVAlNtSmxrb3Ccwrq9qgH1AAV9r+OII1XEMMhVg=="
	          ]
	        }': signature verification failed; please verify account number (0) and chain-id (consumer-localnet): (unable to verify single signer signature): unauthorized

goroutine 48 [running]:
github.com/cosmos/cosmos-sdk/x/genutil.AppModule.InitGenesis({{_}, {_, _}, {_, _}, {_, _}, {_, _}}, {{0x338a830, ...}, ...}, ...)
	/var/cache/go/github.com/atomone-hub/cosmos-sdk@v0.50.14-atomone.1.0.20260107105933-680c2b513b67/x/genutil/module.go:109 +0x15f
github.com/cosmos/cosmos-sdk/types/module.(*Manager).InitGenesis(_, {{0x338a830, 0x4bb2000}, {0x33a3a80, 0xc0029feec0}, {{0x0, 0x0}, {0xc001500828, 0x11}, 0x0, ...}, ...}, ...)
	/var/cache/go/github.com/atomone-hub/cosmos-sdk@v0.50.14-atomone.1.0.20260107105933-680c2b513b67/types/module/module.go:508 +0x859
github.com/atomone-hub/ics-poc-1/testapp/consumer.(*App).InitChainer(_, {{0x338a830, 0x4bb2000}, {0x33a3a80, 0xc0029feec0}, {{0x0, 0x0}, {0xc001500828, 0x11}, 0x0, ...}, ...}, ...)
	/home/julien/projects/atomone-hub/ics-poc-1/testapp/consumer/app.go:584 +0xe5
github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).InitChain(0xc000d44248, 0xc0028752d0)
	/var/cache/go/github.com/atomone-hub/cosmos-sdk@v0.50.14-atomone.1.0.20260107105933-680c2b513b67/baseapp/abci.go:111 +0x722
github.com/cosmos/cosmos-sdk/server.cometABCIWrapper.InitChain(...)
	/var/cache/go/github.com/atomone-hub/cosmos-sdk@v0.50.14-atomone.1.0.20260107105933-680c2b513b67/server/cmt_abci.go:32
github.com/cometbft/cometbft/abci/types._ABCI_InitChain_Handler({0x27f1080, 0xc000c60020}, {0x338abe8, 0xc0029fd650}, 0xc002879700, 0x0)
	/var/cache/go/github.com/atomone-hub/cometbft@v0.0.0-20251216151559-a24cfa28b916/abci/types/types.pb.go:4250 +0x1a6
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000c5e000, {0x338abe8, 0xc000c527b0}, 0xc001202480, 0xc000c52300, 0x49cb1d0, 0x0)
	/var/cache/go/google.golang.org/grpc@v1.75.0/server.go:1431 +0xfe8
google.golang.org/grpc.(*Server).handleStream(0xc000c5e000, {0x338c840, 0xc000c621a0}, 0xc001202480)
	/var/cache/go/google.golang.org/grpc@v1.75.0/server.go:1842 +0xb95
google.golang.org/grpc.(*Server).serveStreams.func2.1()
	/var/cache/go/google.golang.org/grpc@v1.75.0/server.go:1061 +0x7f
created by google.golang.org/grpc.(*Server).serveStreams.func2 in goroutine 47
	/var/cache/go/google.golang.org/grpc@v1.75.0/server.go:1072 +0x11d
make: *** [Makefile:60: consumer-start] Error 2
```

--- a/Makefile
+++ b/Makefile
@@ -33,7 +33,7 @@ provider-start: build
 	# Decrease voting period to 5min
 	jq '.app_state.gov.params.voting_period = "300s"' $(provider_home)/config/genesis.json > /tmp/gen
 	mv /tmp/gen $(provider_home)/config/genesis.json
-	printf "[[chains]]\nchain_id = \"consumer-localnet\"\ngrpc_address = \"tcp://localhost:26658\"\n" > $(provider_home)/config/ics.toml
+	printf "[[chains]]\nchain_id = \"consumer-localnet\"\ngrpc_address = \"tcp://localhost:26658\"\nhome = \"$(HOME)/.consumer-localnet\"\n" > $(provider_home)/config/ics.toml
 	$(providerd) start --rpc.grpc_laddr tcp://127.0.0.1:36658
 
 consumer_home=~/.consumer-localnet
@@ -48,8 +48,6 @@ consumer-start: build
 	$(consumerd) genesis add-genesis-account val 1000000000000uatone --chain-id consumer-localnet
 	$(consumerd) keys add user
 	$(consumerd) genesis add-genesis-account user 1000000000uatone --chain-id consumer-localnet
-	$(consumerd) genesis gentx val 1000000000uatone
-	$(consumerd) genesis collect-gentxs
 
 	# Set validator gas prices
 	sed -i.bak 's#^minimum-gas-prices = .*#minimum-gas-prices = "0.01uatone,0.01uphoton"#g' $(consumer_home)/config/app.toml

--- a/Makefile
+++ b/Makefile
@@ -51,6 +51,8 @@ consumer-start: build
 	$(consumerd) genesis add-genesis-account val 1000000000000uatone --chain-id consumer-localnet
 	$(consumerd) keys add user
 	$(consumerd) genesis add-genesis-account user 1000000000uatone --chain-id consumer-localnet
+	$(consumerd) genesis gentx val 1000000000uatone # todo this should not be needed
+	$(consumerd) genesis collect-gentxs
 
 	# Set validator gas prices
 	sed -i.bak 's#^minimum-gas-prices = .*#minimum-gas-prices = "0.01uatone,0.01uphoton"#g' $(consumer_home)/config/app.toml

--- a/Makefile
+++ b/Makefile
@@ -33,6 +33,9 @@ provider-start: build
 	# Decrease voting period to 5min
 	jq '.app_state.gov.params.voting_period = "300s"' $(provider_home)/config/genesis.json > /tmp/gen
 	mv /tmp/gen $(provider_home)/config/genesis.json
+	# Add consumer chain as active in provider genesis
+	jq '.app_state.provider.consumer_chains = [{"chain_id": "consumer-localnet", "name": "Consumer Localnet", "version": "1.0.0", "status": 2, "added_at": 0, "sunset_at": 0, "module_account_address": ""}]' $(provider_home)/config/genesis.json > /tmp/gen
+	mv /tmp/gen $(provider_home)/config/genesis.json
 	printf "[[chains]]\nchain_id = \"consumer-localnet\"\ngrpc_address = \"tcp://localhost:26658\"\nhome = \"$(HOME)/.consumer-localnet\"\n" > $(provider_home)/config/ics.toml
 	$(providerd) start --rpc.grpc_laddr tcp://127.0.0.1:36658
 

--- a/abci/abci.go
+++ b/abci/abci.go
@@ -51,7 +51,7 @@ func (m *Multiplexer) CheckTx(ctx context.Context, req *abci.RequestCheckTx) (*a
 		return &abci.ResponseCheckTx{Code: 1, Log: fmt.Sprintf("unknown chain: %s", chainID)}, nil
 	}
 
-	if !m.initializedChains[chainID] {
+	if !m.initializedConsumerChains[chainID] {
 		return &abci.ResponseCheckTx{Code: 1, Log: fmt.Sprintf("chain not initialized: %s", chainID)}, nil
 	}
 
@@ -61,28 +61,37 @@ func (m *Multiplexer) CheckTx(ctx context.Context, req *abci.RequestCheckTx) (*a
 func (m *Multiplexer) InitChain(ctx context.Context, req *abci.RequestInitChain) (*abci.ResponseInitChain, error) {
 	m.logger.Debug("InitChain", "chain_id", req.ChainId)
 
+	providerResp, err := m.providerChain.InitChain(req)
+	if err != nil {
+		return nil, err
+	}
+
+	m.providerValidators = providerResp.Validators
+	m.providerConsensusParams = providerResp.ConsensusParams
+
+	response := &abci.ResponseInitChain{
+		ConsensusParams: providerResp.ConsensusParams,
+		Validators:      providerResp.Validators,
+	}
+
+	chainHashes := make(map[string][]byte)
+	chainHashes[m.providerChainID] = providerResp.AppHash
+
 	type result struct {
 		chainID  string
 		response *abci.ResponseInitChain
 		err      error
 	}
 
-	// Include provider chain in the count
-	numChains := len(m.chainHandlers) + 1
-	results := make(chan result, numChains)
+	results := make(chan result, len(m.chainHandlers))
 	var wg sync.WaitGroup
 
-	// Call provider chain first
-	wg.Go(func() {
-		resp, err := m.providerChain.InitChain(req)
-		results <- result{chainID: req.ChainId, response: resp, err: err}
-	})
-
-	// Then call consumer chains
 	for chainID, handler := range m.chainHandlers {
 		wg.Go(func() {
 			consumerReq := *req
 			consumerReq.ChainId = chainID
+			consumerReq.Validators = m.providerValidators
+			consumerReq.ConsensusParams = m.providerConsensusParams
 
 			// Load consumer chain's own genesis state from its home directory
 			appState, err := handler.config.LoadGenesisAppState()
@@ -103,28 +112,14 @@ func (m *Multiplexer) InitChain(ctx context.Context, req *abci.RequestInitChain)
 		close(results)
 	}()
 
-	response := &abci.ResponseInitChain{}
-	chainHashes := make(map[string][]byte)
-
 	for res := range results {
 		if res.err != nil {
-			if res.chainID == m.providerChainID {
-				return nil, res.err
-			}
-			m.logger.Error("InitChain failed for consumer chain, skipping",
-				"chain_id", res.chainID, "error", res.err)
+			m.logger.Error("InitChain failed for consumer chain, skipping", "chain_id", res.chainID, "error", res.err)
 			continue
 		}
 
 		chainHashes[res.chainID] = res.response.AppHash
-
-		m.initializedChains[res.chainID] = true
-
-		// Only use ConsensusParams and Validators from provider chain
-		if res.chainID == m.providerChainID {
-			response.ConsensusParams = res.response.ConsensusParams
-			response.Validators = append(response.Validators, res.response.Validators...)
-		}
+		m.initializedConsumerChains[res.chainID] = true
 	}
 
 	// Combine app hashes in sorted order by chain ID
@@ -163,7 +158,7 @@ func (m *Multiplexer) ProcessProposal(ctx context.Context, req *abci.RequestProc
 				m.logger.Debug("Transaction for unexisting chain in proposal, skipping...", "chain_id", chainID)
 				continue
 			}
-			if !m.initializedChains[chainID] {
+			if !m.initializedConsumerChains[chainID] {
 				m.logger.Debug("Transaction for uninitialized chain in proposal, skipping...", "chain_id", chainID)
 				continue
 			}
@@ -323,7 +318,7 @@ func (m *Multiplexer) FinalizeBlock(ctx context.Context, req *abci.RequestFinali
 
 	// Initialize any new active chains
 	for chainID := range m.activeChains {
-		if err := m.initChainIfNeeded(chainID, req.Height, req.Time); err != nil {
+		if _, err := m.initChainIfNeeded(chainID, req.Height, req.Time); err != nil {
 			m.logger.Error("Failed to initialize chain", "chain_id", chainID, "error", err)
 		}
 	}
@@ -421,7 +416,7 @@ func (m *Multiplexer) Commit(ctx context.Context, req *abci.RequestCommit) (*abc
 
 	for chainID, handler := range m.chainHandlers {
 		// Only commit if the chain has been initialized
-		if !m.initializedChains[chainID] {
+		if !m.initializedConsumerChains[chainID] {
 			m.logger.Debug("Skipping Commit for uninitialized chain", "chain_id", chainID)
 			continue
 		}

--- a/abci/abci.go
+++ b/abci/abci.go
@@ -66,13 +66,12 @@ func (m *Multiplexer) InitChain(ctx context.Context, req *abci.RequestInitChain)
 		return nil, err
 	}
 
-	m.providerValidators = providerResp.Validators
-	m.providerConsensusParams = providerResp.ConsensusParams
-
 	response := &abci.ResponseInitChain{
 		ConsensusParams: providerResp.ConsensusParams,
 		Validators:      providerResp.Validators,
 	}
+	m.providerGenesisConsensusParams = providerResp.ConsensusParams
+	m.providerGenesisValidators = providerResp.Validators
 
 	chainHashes := make(map[string][]byte)
 	chainHashes[m.providerChainID] = providerResp.AppHash
@@ -90,8 +89,8 @@ func (m *Multiplexer) InitChain(ctx context.Context, req *abci.RequestInitChain)
 		wg.Go(func() {
 			consumerReq := *req
 			consumerReq.ChainId = chainID
-			consumerReq.Validators = m.providerValidators
-			consumerReq.ConsensusParams = m.providerConsensusParams
+			consumerReq.Validators = m.providerGenesisValidators
+			consumerReq.ConsensusParams = m.providerGenesisConsensusParams
 
 			// Load consumer chain's own genesis state from its home directory
 			appState, err := handler.config.LoadGenesisAppState()
@@ -144,22 +143,21 @@ func (m *Multiplexer) PrepareProposal(ctx context.Context, req *abci.RequestPrep
 func (m *Multiplexer) ProcessProposal(ctx context.Context, req *abci.RequestProcessProposal) (*abci.ResponseProcessProposal, error) {
 	m.logger.Debug("ProcessProposal", "num_txs", len(req.Txs))
 
+	// Parse and categorize transactions by chain
 	chainTxs := make(map[string][][]byte)
-
 	for _, tx := range req.Txs {
 		chainID, payload, err := ParseHeader(tx)
 		if err != nil {
 			return &abci.ResponseProcessProposal{Status: abci.ResponseProcessProposal_REJECT}, nil
 		}
 
-		// Check if it's provider chain or consumer chain
 		if chainID != m.providerChainID {
 			if _, exists := m.chainHandlers[chainID]; !exists {
-				m.logger.Debug("Transaction for unexisting chain in proposal, skipping...", "chain_id", chainID)
+				m.logger.Debug("Transaction for unknown chain in proposal, skipping", "chain_id", chainID)
 				continue
 			}
 			if !m.initializedConsumerChains[chainID] {
-				m.logger.Debug("Transaction for uninitialized chain in proposal, skipping...", "chain_id", chainID)
+				m.logger.Debug("Transaction for uninitialized chain in proposal, skipping", "chain_id", chainID)
 				continue
 			}
 		}
@@ -184,12 +182,10 @@ func (m *Multiplexer) ProcessProposal(ctx context.Context, req *abci.RequestProc
 			var resp *abci.ResponseProcessProposal
 			var err error
 
-			// Call provider chain or consumer chain
 			if chainID == m.providerChainID {
 				resp, err = m.providerChain.ProcessProposal(&chainReq)
 			} else {
-				handler := m.chainHandlers[chainID]
-				resp, err = handler.app.ProcessProposal(&chainReq)
+				resp, err = m.chainHandlers[chainID].app.ProcessProposal(&chainReq)
 			}
 
 			results <- result{chainID: chainID, response: resp, err: err}
@@ -204,23 +200,20 @@ func (m *Multiplexer) ProcessProposal(ctx context.Context, req *abci.RequestProc
 	// Clear rejected chains from previous proposal
 	m.rejectedConsumerChains = make(map[string]bool)
 
-	// Only reject proposal if provider chain rejects
-	// Consumer chain rejections will skip that chain's block
 	for res := range results {
 		if res.err != nil {
 			return nil, res.err
 		}
 
-		// If provider chain rejects, reject the entire proposal
+		// Provider chain rejection rejects entire proposal
 		if res.chainID == m.providerChainID && res.response.Status != abci.ResponseProcessProposal_ACCEPT {
-			m.logger.Info("Provider chain rejected proposal, rejecting entire proposal")
+			m.logger.Warn("Provider chain rejected proposal, rejecting entire proposal")
 			return &abci.ResponseProcessProposal{Status: res.response.Status}, nil
 		}
 
-		// If consumer chain rejects, mark it to skip block in FinalizeBlock
+		// Consumer chain rejection marks it to skip in FinalizeBlock
 		if res.chainID != m.providerChainID && res.response.Status != abci.ResponseProcessProposal_ACCEPT {
-			m.logger.Info("Consumer chain rejected proposal, will skip block for this chain",
-				"chain_id", res.chainID, "status", res.response.Status)
+			m.logger.Warn("Consumer chain rejected proposal, will skip block", "chain_id", res.chainID, "status", res.response.Status)
 			m.rejectedConsumerChains[res.chainID] = true
 		}
 	}
@@ -231,15 +224,14 @@ func (m *Multiplexer) ProcessProposal(ctx context.Context, req *abci.RequestProc
 func (m *Multiplexer) FinalizeBlock(ctx context.Context, req *abci.RequestFinalizeBlock) (*abci.ResponseFinalizeBlock, error) {
 	m.logger.Info("FinalizeBlock", "height", req.Height, "num_txs", len(req.Txs))
 
-	// Check halt conditions
 	if err := m.checkHaltConditions(req); err != nil {
 		return nil, fmt.Errorf("failed to finalize block because the node should halt: %w", err)
 	}
 
-	// Parse and categorize transactions by chain
+	// parse and categorize transactions by chain
 	chainTxs := make(map[string][][]byte)
 	txPositions := make(map[string][]int)
-	skippedTxIndices := make([]int, 0)
+	var skippedIndices []int
 
 	for idx, tx := range req.Txs {
 		chainID, payload, err := ParseHeader(tx)
@@ -247,82 +239,39 @@ func (m *Multiplexer) FinalizeBlock(ctx context.Context, req *abci.RequestFinali
 			return nil, fmt.Errorf("invalid tx at index %d: %w", idx, err)
 		}
 
-		// Check if it's for provider or consumer chain
-		if chainID == m.providerChainID {
-			chainTxs[m.providerChainID] = append(chainTxs[m.providerChainID], payload)
-			txPositions[m.providerChainID] = append(txPositions[m.providerChainID], idx)
-		} else {
-			// Skip transactions for rejected consumer chains
-			if m.rejectedConsumerChains[chainID] {
-				m.logger.Debug("Skipping transaction for rejected consumer chain",
-					"chain_id", chainID, "tx_index", idx)
-				skippedTxIndices = append(skippedTxIndices, idx)
+		if m.rejectedConsumerChains[chainID] {
+			m.logger.Debug("Skipping transaction for rejected consumer chain", "chain_id", chainID, "tx_index", idx)
+			skippedIndices = append(skippedIndices, idx)
+			continue
+		}
+
+		if chainID != m.providerChainID {
+			if _, exists := m.chainHandlers[chainID]; !exists {
+				m.logger.Debug("Unknown chain for tx at index", "chain_id", chainID, "tx_index", idx)
+				skippedIndices = append(skippedIndices, idx)
 				continue
 			}
-
-			handler, exists := m.chainHandlers[chainID]
-			if !exists {
-				return nil, fmt.Errorf("unknown chain for tx at index %d: %s", idx, chainID)
-			}
-
-			chainTxs[handler.ChainID] = append(chainTxs[handler.ChainID], payload)
-			txPositions[handler.ChainID] = append(txPositions[handler.ChainID], idx)
 		}
+
+		chainTxs[chainID] = append(chainTxs[chainID], payload)
+		txPositions[chainID] = append(txPositions[chainID], idx)
 	}
 
-	response := &abci.ResponseFinalizeBlock{
-		TxResults: make([]*abci.ExecTxResult, len(req.Txs)),
-	}
-
-	// Mark skipped transactions with error results
-	for _, idx := range skippedTxIndices {
-		response.TxResults[idx] = &abci.ExecTxResult{
-			Code: 1,
-			Log:  "transaction skipped: consumer chain rejected proposal",
-		}
-	}
-
-	chainHashes := make(map[string][]byte)
-	var events []abci.Event
-
-	// ALWAYS call FinalizeBlock on provider chain first (even with no transactions)
-	// This ensures provider state is properly set up before Commit
+	// process provider chain first
 	providerReq := *req
-	providerReq.Txs = chainTxs[m.providerChainID] // may be nil/empty, that's ok
+	providerReq.Txs = chainTxs[m.providerChainID]
 	providerResp, err := m.providerChain.FinalizeBlock(&providerReq)
 	if err != nil {
 		return nil, fmt.Errorf("provider chain FinalizeBlock failed: %w", err)
 	}
 
-	// Store provider results
-	chainHashes[m.providerChainID] = providerResp.AppHash
-	events = append(events, providerResp.Events...)
-	response.ConsensusParamUpdates = providerResp.ConsensusParamUpdates
-	response.ValidatorUpdates = append(response.ValidatorUpdates, providerResp.ValidatorUpdates...)
-
-	// Map provider tx results to their original positions
-	for i, pos := range txPositions[m.providerChainID] {
-		if i < len(providerResp.TxResults) {
-			response.TxResults[pos] = providerResp.TxResults[i]
-		}
-	}
-
-	// Now that provider FinalizeBlock is done, we can query for active chains
 	if req.Height > 1 {
 		if err := m.updateActiveChains(ctx); err != nil {
 			m.logger.Error("Failed to update active chains", "error", err)
 		}
 	}
 
-	m.validateActiveChains()
-
-	// Initialize any new active chains
-	for chainID := range m.activeChains {
-		if _, err := m.initChainIfNeeded(chainID, req.Height, req.Time); err != nil {
-			m.logger.Error("Failed to initialize chain", "chain_id", chainID, "error", err)
-		}
-	}
-
+	// process active consumer chains in parallel
 	type result struct {
 		chainID   string
 		response  *abci.ResponseFinalizeBlock
@@ -330,60 +279,84 @@ func (m *Multiplexer) FinalizeBlock(ctx context.Context, req *abci.RequestFinali
 		err       error
 	}
 
-	// Count consumer chains that have transactions
-	numConsumerChains := 0
-	for chainID := range chainTxs {
-		if chainID != m.providerChainID {
-			numConsumerChains++
+	results := make(chan result, len(m.activeChains))
+	var wg sync.WaitGroup
+
+	for chainID := range m.activeChains {
+		if m.rejectedConsumerChains[chainID] {
+			continue
+		}
+
+		handler, exists := m.chainHandlers[chainID]
+		if !exists {
+			m.logger.Error("Active consumer chain missing handler", "chain_id", chainID)
+			continue
+		}
+
+		if _, err := m.initChainIfNeeded(chainID, req.Height, req.Time); err != nil {
+			m.logger.Error("Failed to initialize chain", "chain_id", chainID, "error", err)
+			continue
+		}
+
+		if !m.initializedConsumerChains[chainID] {
+			m.logger.Warn("Skipping uninitialized chain", "chain_id", chainID)
+			continue
+		}
+
+		txs := chainTxs[chainID]
+		positions := txPositions[chainID]
+		wg.Go(func() {
+			chainReq := *req
+			chainReq.Txs = txs
+			resp, err := handler.app.FinalizeBlock(&chainReq)
+			results <- result{chainID: chainID, response: resp, positions: positions, err: err}
+		})
+	}
+
+	go func() {
+		wg.Wait()
+		close(results)
+	}()
+
+	response := &abci.ResponseFinalizeBlock{
+		TxResults:             make([]*abci.ExecTxResult, len(req.Txs)),
+		ConsensusParamUpdates: providerResp.ConsensusParamUpdates,
+		ValidatorUpdates:      providerResp.ValidatorUpdates,
+	}
+
+	chainHashes := map[string][]byte{m.providerChainID: providerResp.AppHash}
+	events := providerResp.Events
+
+	// Map provider tx results
+	for i, pos := range txPositions[m.providerChainID] {
+		if i < len(providerResp.TxResults) {
+			response.TxResults[pos] = providerResp.TxResults[i]
 		}
 	}
 
-	if numConsumerChains > 0 {
-		results := make(chan result, numConsumerChains)
-		var wg sync.WaitGroup
+	// Mark skipped transactions
+	for _, idx := range skippedIndices {
+		response.TxResults[idx] = &abci.ExecTxResult{
+			Code: 1,
+			Log:  "transaction skipped",
+		}
+	}
 
-		for chainID, txs := range chainTxs {
-			if chainID == m.providerChainID {
-				continue // already processed
-			}
-			positions := txPositions[chainID]
-			wg.Go(func() {
-				chainReq := *req
-				chainReq.Txs = txs
-
-				handler := m.chainHandlers[chainID]
-				resp, err := handler.app.FinalizeBlock(&chainReq)
-
-				results <- result{
-					chainID:   chainID,
-					response:  resp,
-					positions: positions,
-					err:       err,
-				}
-			})
+	// collect consumer chain results
+	for res := range results {
+		if res.err != nil {
+			m.logger.Error("FinalizeBlock failed for consumer chain", "chain_id", res.chainID, "error", res.err)
+			continue
 		}
 
-		go func() {
-			wg.Wait()
-			close(results)
-		}()
-
-		for res := range results {
-			if res.err != nil {
-				m.logger.Error("FinalizeBlock failed for consumer chain, skipping block for this chain",
-					"chain_id", res.chainID, "error", res.err)
-				continue
+		for i, pos := range res.positions {
+			if i < len(res.response.TxResults) {
+				response.TxResults[pos] = res.response.TxResults[i]
 			}
-
-			for i, pos := range res.positions {
-				if i < len(res.response.TxResults) {
-					response.TxResults[pos] = res.response.TxResults[i]
-				}
-			}
-
-			chainHashes[res.chainID] = res.response.AppHash
-			events = append(events, res.response.Events...)
 		}
+
+		chainHashes[res.chainID] = res.response.AppHash
+		events = append(events, res.response.Events...)
 	}
 
 	// Combine app hashes in sorted order
@@ -396,13 +369,8 @@ func (m *Multiplexer) FinalizeBlock(ctx context.Context, req *abci.RequestFinali
 	for _, chainID := range sortedChainIDs {
 		response.AppHash = append(response.AppHash, chainHashes[chainID]...)
 	}
-
 	response.Events = events
 
-	// Clear rejected consumer chains after finalizing block
-	m.rejectedConsumerChains = make(map[string]bool)
-
-	m.logger.Info("FinalizeBlock complete", "height", req.Height)
 	return response, nil
 }
 
@@ -414,10 +382,20 @@ func (m *Multiplexer) Commit(ctx context.Context, req *abci.RequestCommit) (*abc
 		return nil, err
 	}
 
-	for chainID, handler := range m.chainHandlers {
-		// Only commit if the chain has been initialized
+	// Only commit active and initialized chains
+	for chainID := range m.activeChains {
 		if !m.initializedConsumerChains[chainID] {
 			m.logger.Debug("Skipping Commit for uninitialized chain", "chain_id", chainID)
+			continue
+		}
+
+		if m.rejectedConsumerChains[chainID] {
+			continue
+		}
+
+		handler, exists := m.chainHandlers[chainID]
+		if !exists {
+			m.logger.Error("Skipping Commit for missing chain handler", "chain_id", chainID)
 			continue
 		}
 

--- a/abci/abci.go
+++ b/abci/abci.go
@@ -84,6 +84,15 @@ func (m *Multiplexer) InitChain(ctx context.Context, req *abci.RequestInitChain)
 			consumerReq := *req
 			consumerReq.ChainId = chainID
 
+			// Load consumer chain's own genesis state from its home directory
+			appState, err := handler.config.LoadGenesisAppState()
+			if err != nil {
+				m.logger.Error("Failed to load genesis for consumer chain", "chain_id", chainID, "error", err)
+				results <- result{chainID: chainID, response: nil, err: err}
+				return
+			}
+			consumerReq.AppStateBytes = appState
+
 			resp, err := handler.app.InitChain(&consumerReq)
 			results <- result{chainID: chainID, response: resp, err: err}
 		})

--- a/abci/multiplexer.go
+++ b/abci/multiplexer.go
@@ -21,9 +21,9 @@ import (
 	"github.com/cometbft/cometbft/node"
 	"github.com/cometbft/cometbft/p2p"
 	pvm "github.com/cometbft/cometbft/privval"
+	cmtproto "github.com/cometbft/cometbft/proto/tendermint/types"
 	"github.com/cometbft/cometbft/proxy"
 	"github.com/cometbft/cometbft/rpc/client/local"
-	cmttypes "github.com/cometbft/cometbft/types"
 	db "github.com/cosmos/cosmos-db"
 	"github.com/cosmos/cosmos-sdk/client"
 	"github.com/cosmos/cosmos-sdk/codec"
@@ -75,14 +75,14 @@ type Multiplexer struct {
 	chainHandlers map[string]*ChainHandler
 	// rejectedConsumerChains tracks consumer chains that rejected the current proposal
 	rejectedConsumerChains map[string]bool
-	// initializedChains tracks which chains have had InitChain called
-	initializedChains map[string]bool
+	// initializedConsumerChains tracks which consumer chains have had InitChain called
+	initializedConsumerChains map[string]bool
 	// activeChains tracks the list of active chains from the provider module
 	activeChains map[string]bool
-	// genesisValidators stores the initial validators from InitChain for use with dynamic consumer chains
-	genesisValidators []abci.ValidatorUpdate
-	// genesisConsensusParams stores the initial consensus params from InitChain for use with dynamic consumer chains
-	genesisConsensusParams *cmttypes.ConsensusParams
+	// providerValidators stores the validators from the provider's InitChain response for use with consumer chains
+	providerValidators []abci.ValidatorUpdate
+	// providerConsensusParams stores the initial consensus params from the provider's InitChain for use with consumer chains
+	providerConsensusParams *cmtproto.ConsensusParams
 	// cmNode is the comet node which has been created (of the provider chain).
 	cmNode *node.Node
 	// ctx is the context which is passed to the comet, grpc and api server starting functions.
@@ -112,16 +112,16 @@ func NewMultiplexer(
 	}
 
 	mp := &Multiplexer{
-		svrCtx:                 svrCtx,
-		svrCfg:                 svrCfg,
-		clientContext:          clientCtx,
-		providerCreator:        providerCreator,
-		logger:                 logger,
-		chainHandlers:          make(map[string]*ChainHandler, len(chainConfig.Chains)),
-		rejectedConsumerChains: make(map[string]bool),
-		initializedChains:      make(map[string]bool),
-		activeChains:           make(map[string]bool),
-		chainConfig:            chainConfig,
+		svrCtx:                    svrCtx,
+		svrCfg:                    svrCfg,
+		clientContext:             clientCtx,
+		providerCreator:           providerCreator,
+		logger:                    logger,
+		chainHandlers:             make(map[string]*ChainHandler, len(chainConfig.Chains)),
+		rejectedConsumerChains:    make(map[string]bool),
+		initializedConsumerChains: make(map[string]bool),
+		activeChains:              make(map[string]bool),
+		chainConfig:               chainConfig,
 	}
 
 	return mp, nil
@@ -191,26 +191,34 @@ func (m *Multiplexer) startNativeProvider() error {
 
 // loadGenesisValues loads validators and consensus params from the genesis document.
 func (m *Multiplexer) loadGenesisValues() error {
+	// skip fetching if we have data
+	if len(m.providerValidators) != 0 && m.providerConsensusParams != nil {
+		return nil
+	}
+
 	genDocProvider := GetGenDocProvider(m.svrCtx.Config)
 	genDoc, err := genDocProvider()
 	if err != nil {
 		return fmt.Errorf("failed to load genesis doc: %w", err)
 	}
 
-	// Convert genesis validators to ABCI validator updates
-	m.genesisValidators = make([]abci.ValidatorUpdate, 0, len(genDoc.Validators))
+	// Convert genesis validators to ABCI validator updates as fallback
+	// These will be overwritten by the provider's InitChain response validators
+	m.providerValidators = make([]abci.ValidatorUpdate, 0, len(genDoc.Validators))
 	for _, val := range genDoc.Validators {
 		pubKey, err := encoding.PubKeyToProto(val.PubKey)
 		if err != nil {
 			return fmt.Errorf("failed to convert validator pubkey: %w", err)
 		}
-		m.genesisValidators = append(m.genesisValidators, abci.ValidatorUpdate{
+		m.providerValidators = append(m.providerValidators, abci.ValidatorUpdate{
 			PubKey: pubKey,
 			Power:  val.Power,
 		})
 	}
 
-	m.genesisConsensusParams = genDoc.ConsensusParams
+	consensusParams := genDoc.ConsensusParams.ToProto()
+	m.providerConsensusParams = &consensusParams
+
 	return nil
 }
 
@@ -531,57 +539,48 @@ func (m *Multiplexer) validateActiveChains() {
 
 // initChainIfNeeded initializes a chain if it hasn't been initialized yet.
 // Checks chain state via Info to detect already-initialized chains after restart.
-func (m *Multiplexer) initChainIfNeeded(chainID string, height int64, blockTime time.Time) error {
-	if m.initializedChains[chainID] {
-		return nil
+func (m *Multiplexer) initChainIfNeeded(chainID string, height int64, blockTime time.Time) ([]byte, error) {
+	if m.initializedConsumerChains[chainID] {
+		return nil, nil
 	}
 
 	handler, exists := m.chainHandlers[chainID]
 	if !exists {
-		return fmt.Errorf("chain handler not found for chain_id: %s", chainID)
+		return nil, fmt.Errorf("chain handler not found for chain_id: %s", chainID)
 	}
 
 	// Check if chain has state (already initialized)
 	infoResp, err := handler.app.Info(&abci.RequestInfo{})
 	if err == nil && (infoResp.LastBlockHeight > 0 || len(infoResp.LastBlockAppHash) > 0) {
 		m.logger.Info("Chain already initialized", "chain_id", chainID, "height", infoResp.LastBlockHeight)
-		m.initializedChains[chainID] = true
-		return nil
+		m.initializedConsumerChains[chainID] = true
+		return nil, nil
 	}
 
 	m.logger.Info("Initializing new consumer chain", "chain_id", chainID, "height", height)
 
-	// Use stored genesis values, falling back to defaults if not available
-	consensusParams := m.genesisConsensusParams
-	if consensusParams == nil {
-		consensusParams = cmttypes.DefaultConsensusParams()
-	}
-
 	// Load consumer chain's own genesis state from its home directory
 	appState, err := handler.config.LoadGenesisAppState()
 	if err != nil {
-		return fmt.Errorf("failed to load genesis for chain %s: %w", chainID, err)
+		return nil, fmt.Errorf("failed to load genesis for chain %s: %w", chainID, err)
 	}
 
-	protoConsensusParams := consensusParams.ToProto()
 	initReq := &abci.RequestInitChain{
 		Time:            blockTime,
 		ChainId:         chainID,
-		ConsensusParams: &protoConsensusParams,
-		Validators:      m.genesisValidators,
+		ConsensusParams: m.providerConsensusParams,
+		Validators:      m.providerValidators,
 		AppStateBytes:   appState,
 		InitialHeight:   height,
 	}
 
-	_, err = handler.app.InitChain(initReq)
+	resp, err := handler.app.InitChain(initReq)
 	if err != nil {
-		return fmt.Errorf("failed to initialize chain %s: %w", chainID, err)
+		return nil, fmt.Errorf("failed to initialize chain %s: %w", chainID, err)
 	}
+	m.initializedConsumerChains[chainID] = true
 
-	m.initializedChains[chainID] = true
-	m.logger.Info("Successfully initialized consumer chain", "chain_id", chainID)
-
-	return nil
+	return resp.AppHash, nil
 }
 
 // updateActiveChains updates the list of active chains from the provider module

--- a/abci/multiplexer.go
+++ b/abci/multiplexer.go
@@ -53,6 +53,7 @@ type ChainHandler struct {
 	ChainID string
 	conn    *grpc.ClientConn
 	app     *RemoteABCIClient
+	config  config.ChainInfo
 }
 
 // Multiplexer implements ABCI Application interface and routes requests to multiple chain applications
@@ -223,7 +224,9 @@ func (m *Multiplexer) initChainHandlers() error {
 
 		handler := &ChainHandler{
 			ChainID: chainInfo.ChainID,
+			conn:    chainConn,
 			app:     NewRemoteABCIClient(chainConn),
+			config:  chainInfo,
 		}
 
 		if _, exists := m.chainHandlers[chainInfo.ChainID]; exists {
@@ -554,13 +557,19 @@ func (m *Multiplexer) initChainIfNeeded(chainID string, height int64, blockTime 
 		consensusParams = cmttypes.DefaultConsensusParams()
 	}
 
+	// Load consumer chain's own genesis state from its home directory
+	appState, err := handler.config.LoadGenesisAppState()
+	if err != nil {
+		return fmt.Errorf("failed to load genesis for chain %s: %w", chainID, err)
+	}
+
 	protoConsensusParams := consensusParams.ToProto()
 	initReq := &abci.RequestInitChain{
 		Time:            blockTime,
 		ChainId:         chainID,
 		ConsensusParams: &protoConsensusParams,
 		Validators:      m.genesisValidators,
-		AppStateBytes:   []byte{},
+		AppStateBytes:   appState,
 		InitialHeight:   height,
 	}
 

--- a/abci/multiplexer.go
+++ b/abci/multiplexer.go
@@ -79,10 +79,10 @@ type Multiplexer struct {
 	initializedConsumerChains map[string]bool
 	// activeChains tracks the list of active chains from the provider module
 	activeChains map[string]bool
-	// providerValidators stores the validators from the provider's InitChain response for use with consumer chains
-	providerValidators []abci.ValidatorUpdate
-	// providerConsensusParams stores the initial consensus params from the provider's InitChain for use with consumer chains
-	providerConsensusParams *cmtproto.ConsensusParams
+	// providerGenesisValidators stores the validators from the provider's InitChain response for use with consumer chains
+	providerGenesisValidators []abci.ValidatorUpdate
+	// providerGenesisConsensusParams stores the initial consensus params from the provider's InitChain for use with consumer chains
+	providerGenesisConsensusParams *cmtproto.ConsensusParams
 	// cmNode is the comet node which has been created (of the provider chain).
 	cmNode *node.Node
 	// ctx is the context which is passed to the comet, grpc and api server starting functions.
@@ -192,7 +192,7 @@ func (m *Multiplexer) startNativeProvider() error {
 // loadGenesisValues loads validators and consensus params from the genesis document.
 func (m *Multiplexer) loadGenesisValues() error {
 	// skip fetching if we have data
-	if len(m.providerValidators) != 0 && m.providerConsensusParams != nil {
+	if len(m.providerGenesisValidators) != 0 && m.providerGenesisConsensusParams != nil {
 		return nil
 	}
 
@@ -204,20 +204,20 @@ func (m *Multiplexer) loadGenesisValues() error {
 
 	// Convert genesis validators to ABCI validator updates as fallback
 	// These will be overwritten by the provider's InitChain response validators
-	m.providerValidators = make([]abci.ValidatorUpdate, 0, len(genDoc.Validators))
+	m.providerGenesisValidators = make([]abci.ValidatorUpdate, 0, len(genDoc.Validators))
 	for _, val := range genDoc.Validators {
 		pubKey, err := encoding.PubKeyToProto(val.PubKey)
 		if err != nil {
 			return fmt.Errorf("failed to convert validator pubkey: %w", err)
 		}
-		m.providerValidators = append(m.providerValidators, abci.ValidatorUpdate{
+		m.providerGenesisValidators = append(m.providerGenesisValidators, abci.ValidatorUpdate{
 			PubKey: pubKey,
 			Power:  val.Power,
 		})
 	}
 
 	consensusParams := genDoc.ConsensusParams.ToProto()
-	m.providerConsensusParams = &consensusParams
+	m.providerGenesisConsensusParams = &consensusParams
 
 	return nil
 }
@@ -524,19 +524,6 @@ func (m *Multiplexer) getActiveChainIDsFromProvider(ctx context.Context) ([]stri
 	return activeChainIDs, nil
 }
 
-// validateActiveChains checks that all active chains have corresponding handlers
-func (m *Multiplexer) validateActiveChains() {
-	for chainID := range m.activeChains {
-		if _, exists := m.chainHandlers[chainID]; !exists {
-			m.logger.Error(
-				"Active consumer chain missing from validator configuration",
-				"chain_id", chainID,
-				"error", "chain not found in chainHandlers",
-			)
-		}
-	}
-}
-
 // initChainIfNeeded initializes a chain if it hasn't been initialized yet.
 // Checks chain state via Info to detect already-initialized chains after restart.
 func (m *Multiplexer) initChainIfNeeded(chainID string, height int64, blockTime time.Time) ([]byte, error) {
@@ -568,8 +555,8 @@ func (m *Multiplexer) initChainIfNeeded(chainID string, height int64, blockTime 
 	initReq := &abci.RequestInitChain{
 		Time:            blockTime,
 		ChainId:         chainID,
-		ConsensusParams: m.providerConsensusParams,
-		Validators:      m.providerValidators,
+		ConsensusParams: m.providerGenesisConsensusParams,
+		Validators:      m.providerGenesisValidators,
 		AppStateBytes:   appState,
 		InitialHeight:   height,
 	}

--- a/config/config.go
+++ b/config/config.go
@@ -6,6 +6,8 @@ import (
 	"path/filepath"
 
 	"github.com/spf13/viper"
+
+	genutiltypes "github.com/cosmos/cosmos-sdk/x/genutil/types"
 )
 
 // Config represents configuration for consumer chains
@@ -17,6 +19,22 @@ type Config struct {
 type ChainInfo struct {
 	ChainID     string `mapstructure:"chain_id"`
 	GRPCAddress string `mapstructure:"grpc_address"`
+	Home        string `mapstructure:"home"`
+}
+
+// GenesisFile returns the path to the genesis file for the chain
+func (c *ChainInfo) GenesisFile() string {
+	return filepath.Join(c.Home, "config", "genesis.json")
+}
+
+// LoadGenesisAppState loads and returns the app state bytes from the genesis file
+func (c *ChainInfo) LoadGenesisAppState() ([]byte, error) {
+	genesisFile := c.GenesisFile()
+	appGenesis, err := genutiltypes.AppGenesisFromFile(genesisFile)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load genesis from %s: %w", genesisFile, err)
+	}
+	return appGenesis.AppState, nil
 }
 
 func (c *Config) Validate() error {
@@ -26,6 +44,9 @@ func (c *Config) Validate() error {
 		}
 		if app.GRPCAddress == "" {
 			return fmt.Errorf("address required for %s", app.ChainID)
+		}
+		if app.Home == "" {
+			return fmt.Errorf("home required for %s", app.ChainID)
 		}
 	}
 

--- a/config/config_test.go
+++ b/config/config_test.go
@@ -29,6 +29,9 @@ func TestLoadConfig_WithValidFile(t *testing.T) {
 	if cfg.Chains[0].GRPCAddress != "tcp://localhost:9090" {
 		t.Errorf("expected address 'tcp://localhost:9090', got '%s'", cfg.Chains[0].GRPCAddress)
 	}
+	if cfg.Chains[0].Home != "/tmp/chain1" {
+		t.Errorf("expected home '/tmp/chain1', got '%s'", cfg.Chains[0].Home)
+	}
 
 	// Verify second chain
 	if cfg.Chains[1].ChainID != "chain-2" {
@@ -36,6 +39,9 @@ func TestLoadConfig_WithValidFile(t *testing.T) {
 	}
 	if cfg.Chains[1].GRPCAddress != "tcp://localhost:9091" {
 		t.Errorf("expected address 'tcp://localhost:9091', got '%s'", cfg.Chains[1].GRPCAddress)
+	}
+	if cfg.Chains[1].Home != "/tmp/chain2" {
+		t.Errorf("expected home '/tmp/chain2', got '%s'", cfg.Chains[1].Home)
 	}
 }
 
@@ -82,6 +88,7 @@ func TestValidate_Success(t *testing.T) {
 			{
 				ChainID:     "test-chain",
 				GRPCAddress: "grpc://localhost:9090",
+				Home:        "/home/test/.test-chain",
 			},
 		},
 	}
@@ -98,6 +105,7 @@ func TestValidate_MissingChainID(t *testing.T) {
 			{
 				ChainID:     "",
 				GRPCAddress: "grpc://localhost:9090",
+				Home:        "/home/test/.test-chain",
 			},
 		},
 	}
@@ -114,6 +122,7 @@ func TestValidate_MissingAddress(t *testing.T) {
 			{
 				ChainID:     "test-chain",
 				GRPCAddress: "",
+				Home:        "/home/test/.test-chain",
 			},
 		},
 	}
@@ -124,16 +133,35 @@ func TestValidate_MissingAddress(t *testing.T) {
 	}
 }
 
+func TestValidate_MissingHome(t *testing.T) {
+	cfg := &Config{
+		Chains: []ChainInfo{
+			{
+				ChainID:     "test-chain",
+				GRPCAddress: "grpc://localhost:9090",
+				Home:        "",
+			},
+		},
+	}
+
+	err := cfg.Validate()
+	if err == nil {
+		t.Fatal("expected validation error for missing home, got nil")
+	}
+}
+
 func TestValidate_MultipleChains(t *testing.T) {
 	cfg := &Config{
 		Chains: []ChainInfo{
 			{
 				ChainID:     "chain-1",
 				GRPCAddress: "grpc://localhost:9090",
+				Home:        "/home/test/.chain-1",
 			},
 			{
 				ChainID:     "chain-2",
 				GRPCAddress: "grpc://localhost:9091",
+				Home:        "/home/test/.chain-2",
 			},
 		},
 	}

--- a/modules/consumer/README.md
+++ b/modules/consumer/README.md
@@ -10,14 +10,13 @@ This folder contains a set of module consumer chains are required to use or shou
 ### To use
 
 - modules/consumer/staking: A wrapper of Cosmos SDK x/staking for consumer chains to not handle validator set changes and slashing.
-- TBD.
-  - modules/consumer/consensus: A wrapper for consumer chains to only be able to query consensus params.
 
 ### To not use
 
 - x/upgrade: Upgrade handling is done via the provider chain governance to simplify the life of validators. Additionally, the x/upgrade module rely on the consensus params app version, that is shared with the provider chain. The module cannot effectively work.
 - x/evidence: Evidence handling is done via the provider chain, as a mishevior on the consumer chain cannot happen without a misbehavior on the provider chain. (ref: https://github.com/atomone-hub/ics-poc-1/issues/2)
 - x/slashing: Slashing is handled via the provider chain for the same reason as x/evidence.
+- x/consensus: The chain inherits the consensus from the provider, the x/consensus module is not required.
 
 ## Recommended
 

--- a/modules/provider/keeper/genesis.go
+++ b/modules/provider/keeper/genesis.go
@@ -2,6 +2,7 @@ package keeper
 
 import (
 	"context"
+	"fmt"
 
 	"github.com/atomone-hub/ics-poc-1/modules/provider/types"
 )
@@ -14,6 +15,17 @@ func (k Keeper) InitGenesis(ctx context.Context, genState types.GenesisState) er
 
 	// Initialize consumer chains
 	for _, consumer := range genState.ConsumerChains {
+		if consumer.ModuleAccountAddress != "" {
+			return fmt.Errorf("module account address for %s must be left empty at init genesis", consumer.ChainId)
+		}
+
+		moduleAccount, err := k.CreateConsumerAccount(ctx, consumer.ChainId)
+		if err != nil {
+			return fmt.Errorf("failed to create module account for %s: %w", consumer.ChainId, err)
+		}
+
+		consumer.ModuleAccountAddress = moduleAccount
+
 		if err := k.ConsumerChains.Set(ctx, consumer.ChainId, consumer); err != nil {
 			return err
 		}

--- a/modules/provider/keeper/keeper.go
+++ b/modules/provider/keeper/keeper.go
@@ -10,6 +10,7 @@ import (
 	"cosmossdk.io/math"
 	"github.com/cosmos/cosmos-sdk/codec"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
 
 	"github.com/atomone-hub/ics-poc-1/modules/provider/types"
 )
@@ -90,23 +91,17 @@ func (k Keeper) GetAllConsumerChains(ctx context.Context) ([]types.ConsumerChain
 	return consumers, err
 }
 
-// CreateConsumerModuleAccount creates a module account for a consumer chain.
-func (k Keeper) CreateConsumerModuleAccount(ctx context.Context, chainID string) (string, error) {
+// CreateConsumerAccount creates a module account for a consumer chain.
+func (k Keeper) CreateConsumerAccount(ctx context.Context, chainID string) (string, error) {
 	accountName := types.GetConsumerModuleAccountName(chainID)
+	addr := authtypes.NewModuleAddress(accountName)
 
 	// Check if account already exists
-	addr := k.authKeeper.GetModuleAddress(accountName)
-	if addr != nil {
-		// Account already exists, return its address
-		addrStr, err := k.addressCodec.BytesToString(addr)
-		if err != nil {
-			return "", err
-		}
-		return addrStr, nil
-	}
+	if acc := k.authKeeper.GetAccount(ctx, addr); acc == nil {
+		// Create the module account
+		k.authKeeper.SetAccount(ctx, k.authKeeper.NewAccountWithAddress(ctx, addr))
 
-	// Create the module account
-	k.authKeeper.SetAccount(ctx, k.authKeeper.NewAccountWithAddress(ctx, addr))
+	}
 
 	addrStr, err := k.addressCodec.BytesToString(addr)
 	if err != nil {
@@ -128,7 +123,7 @@ func (k Keeper) CollectFeesFromConsumers(ctx context.Context, feesPerBlock sdk.C
 		// Get consumer module account address
 		consumerAddr, err := k.addressCodec.StringToBytes(consumer.ModuleAccountAddress)
 		if err != nil {
-			return true, fmt.Errorf("failed to parse module account address for chain %s: %w", consumer.ChainId, err)
+			return true, fmt.Errorf("failed to parse module account address (%s) for chain %s: %w", consumer.ModuleAccountAddress, consumer.ChainId, err)
 		}
 
 		// Check if account has sufficient balance

--- a/modules/provider/keeper/keeper_test.go
+++ b/modules/provider/keeper/keeper_test.go
@@ -111,7 +111,7 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 				f.authKeeper.EXPECT().SetAccount(gomock.Any(), gomock.Any()).AnyTimes()
 
 				// Create consumer module account
-				consumerAddr, err := f.keeper.CreateConsumerModuleAccount(f.ctx, consumer.ChainId)
+				consumerAddr, err := f.keeper.CreateConsumerAccount(f.ctx, consumer.ChainId)
 				require.NoError(t, err)
 				consumer.ModuleAccountAddress = consumerAddr
 
@@ -144,7 +144,7 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), moduleAddr).Return(authtypes.NewBaseAccountWithAddress(moduleAddr)).AnyTimes()
 				f.authKeeper.EXPECT().SetAccount(gomock.Any(), gomock.Any()).AnyTimes()
 
-				consumerAddr, err := f.keeper.CreateConsumerModuleAccount(f.ctx, consumer.ChainId)
+				consumerAddr, err := f.keeper.CreateConsumerAccount(f.ctx, consumer.ChainId)
 				require.NoError(t, err)
 				consumer.ModuleAccountAddress = consumerAddr
 				require.NoError(t, f.keeper.ConsumerChains.Set(f.ctx, consumer.ChainId, consumer))
@@ -170,7 +170,7 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), moduleAddr).Return(authtypes.NewBaseAccountWithAddress(moduleAddr)).AnyTimes()
 				f.authKeeper.EXPECT().SetAccount(gomock.Any(), gomock.Any()).AnyTimes()
 
-				consumerAddr, err := f.keeper.CreateConsumerModuleAccount(f.ctx, consumer.ChainId)
+				consumerAddr, err := f.keeper.CreateConsumerAccount(f.ctx, consumer.ChainId)
 				require.NoError(t, err)
 				consumer.ModuleAccountAddress = consumerAddr
 				require.NoError(t, f.keeper.ConsumerChains.Set(f.ctx, consumer.ChainId, consumer))
@@ -187,7 +187,7 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), moduleAddr1).Return(authtypes.NewBaseAccountWithAddress(moduleAddr1)).AnyTimes()
 				f.authKeeper.EXPECT().SetAccount(gomock.Any(), gomock.Any()).AnyTimes()
 
-				consumerAddr1, err := f.keeper.CreateConsumerModuleAccount(f.ctx, consumer1.ChainId)
+				consumerAddr1, err := f.keeper.CreateConsumerAccount(f.ctx, consumer1.ChainId)
 				require.NoError(t, err)
 				consumer1.ModuleAccountAddress = consumerAddr1
 				require.NoError(t, f.keeper.ConsumerChains.Set(f.ctx, consumer1.ChainId, consumer1))
@@ -219,7 +219,7 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 				f.authKeeper.EXPECT().GetModuleAddress("consumer_consumer-2").Return(moduleAddr2).AnyTimes()
 				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), moduleAddr2).Return(authtypes.NewBaseAccountWithAddress(moduleAddr2)).AnyTimes()
 
-				consumerAddr2, err := f.keeper.CreateConsumerModuleAccount(f.ctx, consumer2.ChainId)
+				consumerAddr2, err := f.keeper.CreateConsumerAccount(f.ctx, consumer2.ChainId)
 				require.NoError(t, err)
 				consumer2.ModuleAccountAddress = consumerAddr2
 				require.NoError(t, f.keeper.ConsumerChains.Set(f.ctx, consumer2.ChainId, consumer2))
@@ -253,7 +253,7 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), moduleAddr).Return(authtypes.NewBaseAccountWithAddress(moduleAddr)).AnyTimes()
 				f.authKeeper.EXPECT().SetAccount(gomock.Any(), gomock.Any()).AnyTimes()
 
-				consumerAddr, err := f.keeper.CreateConsumerModuleAccount(f.ctx, consumer.ChainId)
+				consumerAddr, err := f.keeper.CreateConsumerAccount(f.ctx, consumer.ChainId)
 				require.NoError(t, err)
 				consumer.ModuleAccountAddress = consumerAddr
 				require.NoError(t, f.keeper.ConsumerChains.Set(f.ctx, consumer.ChainId, consumer))

--- a/modules/provider/keeper/keeper_test.go
+++ b/modules/provider/keeper/keeper_test.go
@@ -104,10 +104,9 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 			feesPerBlock: sdk.NewCoin(denom, math.NewInt(1000)),
 			setupConsumers: func(t *testing.T, f *fixture) {
 				consumer := DefaultConsumer
-				moduleAddr := defaultModuleAddr
 				// Mock auth keeper for module account creation
-				f.authKeeper.EXPECT().GetModuleAddress("consumer_consumer-1").Return(moduleAddr).AnyTimes()
-				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), moduleAddr).Return(authtypes.NewBaseAccountWithAddress(moduleAddr)).AnyTimes()
+				f.authKeeper.EXPECT().GetAccount(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), gomock.Any()).Return(authtypes.NewBaseAccountWithAddress(defaultModuleAddr)).AnyTimes()
 				f.authKeeper.EXPECT().SetAccount(gomock.Any(), gomock.Any()).AnyTimes()
 
 				// Create consumer module account
@@ -138,10 +137,9 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 			feesPerBlock: sdk.NewCoin(denom, math.NewInt(1000)),
 			setupConsumers: func(t *testing.T, f *fixture) {
 				consumer := DefaultConsumer
-				moduleAddr := defaultModuleAddr
 
-				f.authKeeper.EXPECT().GetModuleAddress("consumer_consumer-1").Return(moduleAddr).AnyTimes()
-				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), moduleAddr).Return(authtypes.NewBaseAccountWithAddress(moduleAddr)).AnyTimes()
+				f.authKeeper.EXPECT().GetAccount(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), gomock.Any()).Return(authtypes.NewBaseAccountWithAddress(defaultModuleAddr)).AnyTimes()
 				f.authKeeper.EXPECT().SetAccount(gomock.Any(), gomock.Any()).AnyTimes()
 
 				consumerAddr, err := f.keeper.CreateConsumerAccount(f.ctx, consumer.ChainId)
@@ -163,11 +161,10 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 			feesPerBlock: sdk.NewCoin(denom, math.NewInt(1000)),
 			setupConsumers: func(t *testing.T, f *fixture) {
 				consumer := DefaultConsumer
-				moduleAddr := defaultModuleAddr
 				// Create inactive consumer (PENDING status)
 				consumer.Status = types.ConsumerStatus_CONSUMER_STATUS_PENDING
-				f.authKeeper.EXPECT().GetModuleAddress("consumer_consumer-1").Return(moduleAddr).AnyTimes()
-				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), moduleAddr).Return(authtypes.NewBaseAccountWithAddress(moduleAddr)).AnyTimes()
+				f.authKeeper.EXPECT().GetAccount(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), gomock.Any()).Return(authtypes.NewBaseAccountWithAddress(defaultModuleAddr)).AnyTimes()
 				f.authKeeper.EXPECT().SetAccount(gomock.Any(), gomock.Any()).AnyTimes()
 
 				consumerAddr, err := f.keeper.CreateConsumerAccount(f.ctx, consumer.ChainId)
@@ -182,9 +179,8 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 			feesPerBlock: sdk.NewCoin(denom, math.NewInt(500)),
 			setupConsumers: func(t *testing.T, f *fixture) {
 				consumer1 := DefaultConsumer
-				moduleAddr1 := defaultModuleAddr
-				f.authKeeper.EXPECT().GetModuleAddress("consumer_consumer-1").Return(moduleAddr1).AnyTimes()
-				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), moduleAddr1).Return(authtypes.NewBaseAccountWithAddress(moduleAddr1)).AnyTimes()
+				f.authKeeper.EXPECT().GetAccount(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), gomock.Any()).Return(authtypes.NewBaseAccountWithAddress(defaultModuleAddr)).AnyTimes()
 				f.authKeeper.EXPECT().SetAccount(gomock.Any(), gomock.Any()).AnyTimes()
 
 				consumerAddr1, err := f.keeper.CreateConsumerAccount(f.ctx, consumer1.ChainId)
@@ -215,10 +211,6 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 					"",
 				)
 
-				moduleAddr2 := sdk.AccAddress("consumer-2-module-account")
-				f.authKeeper.EXPECT().GetModuleAddress("consumer_consumer-2").Return(moduleAddr2).AnyTimes()
-				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), moduleAddr2).Return(authtypes.NewBaseAccountWithAddress(moduleAddr2)).AnyTimes()
-
 				consumerAddr2, err := f.keeper.CreateConsumerAccount(f.ctx, consumer2.ChainId)
 				require.NoError(t, err)
 				consumer2.ModuleAccountAddress = consumerAddr2
@@ -248,9 +240,8 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 			feesPerBlock: sdk.NewCoin(denom, math.NewInt(1000)),
 			setupConsumers: func(t *testing.T, f *fixture) {
 				consumer := DefaultConsumer
-				moduleAddr := defaultModuleAddr
-				f.authKeeper.EXPECT().GetModuleAddress("consumer_consumer-1").Return(moduleAddr).AnyTimes()
-				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), moduleAddr).Return(authtypes.NewBaseAccountWithAddress(moduleAddr)).AnyTimes()
+				f.authKeeper.EXPECT().GetAccount(gomock.Any(), gomock.Any()).Return(nil).AnyTimes()
+				f.authKeeper.EXPECT().NewAccountWithAddress(gomock.Any(), gomock.Any()).Return(authtypes.NewBaseAccountWithAddress(defaultModuleAddr)).AnyTimes()
 				f.authKeeper.EXPECT().SetAccount(gomock.Any(), gomock.Any()).AnyTimes()
 
 				consumerAddr, err := f.keeper.CreateConsumerAccount(f.ctx, consumer.ChainId)
@@ -282,7 +273,7 @@ func TestCollectFeesFromConsumers(t *testing.T) {
 				require.NoError(t, f.keeper.ConsumerChains.Set(f.ctx, consumer.ChainId, consumer))
 			},
 			expectedTotalFees: sdk.NewCoin(denom, math.ZeroInt()),
-			error:             fmt.Errorf("failed to parse module account address for chain consumer-1: decoding bech32 failed: invalid separator index -1"),
+			error:             fmt.Errorf("failed to parse module account address (invalid-address) for chain consumer-1: decoding bech32 failed: invalid separator index -1"),
 		},
 	}
 

--- a/modules/provider/keeper/msg_server.go
+++ b/modules/provider/keeper/msg_server.go
@@ -81,7 +81,7 @@ func (k msgServer) AddConsumer(ctx context.Context, req *types.MsgAddConsumer) (
 	}
 
 	// Create module account for this consumer
-	moduleAccountAddr, err := k.CreateConsumerModuleAccount(ctx, req.ChainId)
+	moduleAccountAddr, err := k.CreateConsumerAccount(ctx, req.ChainId)
 	if err != nil {
 		return nil, errorsmod.Wrap(err, "failed to create module account")
 	}

--- a/modules/provider/types/types.go
+++ b/modules/provider/types/types.go
@@ -1,1 +1,0 @@
-package types

--- a/testapp/consumer/app.go
+++ b/testapp/consumer/app.go
@@ -55,9 +55,6 @@ import (
 	"github.com/cosmos/cosmos-sdk/x/bank"
 	bankkeeper "github.com/cosmos/cosmos-sdk/x/bank/keeper"
 	banktypes "github.com/cosmos/cosmos-sdk/x/bank/types"
-	"github.com/cosmos/cosmos-sdk/x/consensus"
-	consensusparamkeeper "github.com/cosmos/cosmos-sdk/x/consensus/keeper"
-	consensusparamtypes "github.com/cosmos/cosmos-sdk/x/consensus/types"
 	distr "github.com/cosmos/cosmos-sdk/x/distribution"
 	distrkeeper "github.com/cosmos/cosmos-sdk/x/distribution/keeper"
 	distrtypes "github.com/cosmos/cosmos-sdk/x/distribution/types"
@@ -106,7 +103,6 @@ var (
 		auth.AppModuleBasic{},
 		vesting.AppModuleBasic{},
 		bank.AppModuleBasic{},
-		consensus.AppModuleBasic{},
 		gov.NewAppModuleBasic(
 			[]govclient.ProposalHandler{
 				paramsclient.ProposalHandler,
@@ -162,9 +158,8 @@ type App struct { // nolint: golint
 	// different fee-pool from the consumer chain ConsumerKeeper
 	DistrKeeper distrkeeper.Keeper
 
-	GovKeeper             *govkeeper.Keeper // Gov Keeper must be a pointer in the app, so we can SetRouter on it correctly
-	ParamsKeeper          paramskeeper.Keeper
-	ConsensusParamsKeeper consensusparamkeeper.Keeper
+	GovKeeper    *govkeeper.Keeper // Gov Keeper must be a pointer in the app, so we can SetRouter on it correctly
+	ParamsKeeper paramskeeper.Keeper
 
 	// the module manager
 	MM *module.Manager
@@ -229,7 +224,6 @@ func New(
 		authtypes.StoreKey, banktypes.StoreKey, stakingtypes.StoreKey,
 		minttypes.StoreKey, distrtypes.StoreKey,
 		govtypes.StoreKey, paramstypes.StoreKey,
-		consensusparamtypes.StoreKey,
 	)
 
 	// register streaming services
@@ -256,16 +250,9 @@ func New(
 		tkeys[paramstypes.TStoreKey],
 	)
 
-	// set the BaseApp's parameter store
-	// upgradetypes.StoreKey -> maybe consensusparamtypes.StoreKey (package consensusparamtypes ("github.com/cosmos/cosmos-sdk/x/consensus/types")
-	app.ConsensusParamsKeeper = consensusparamkeeper.NewKeeper(
-		appCodec,
-		runtime.NewKVStoreService(keys[consensusparamtypes.StoreKey]),
-		authtypes.NewModuleAddress(govtypes.ModuleName).String(),
-		runtime.EventService{},
-	)
-
-	bApp.SetParamStore(&app.ConsensusParamsKeeper.ParamsStore)
+	// disable param store on baseapp, as consensus is handled by provider
+	// TODO: verify if we don't want to share some values
+	bApp.SetParamStore(nil)
 
 	// add keepers
 	app.AccountKeeper = authkeeper.NewAccountKeeper(
@@ -364,12 +351,10 @@ func New(
 		auth.NewAppModule(appCodec, app.AccountKeeper, nil, app.GetSubspace(authtypes.ModuleName)),
 		vesting.NewAppModule(app.AccountKeeper, app.BankKeeper),
 		bank.NewAppModule(appCodec, app.BankKeeper, app.AccountKeeper, app.GetSubspace(banktypes.ModuleName)),
-		consensus.NewAppModule(appCodec, app.ConsensusParamsKeeper),
 		gov.NewAppModule(appCodec, app.GovKeeper, app.AccountKeeper, app.BankKeeper, app.GetSubspace(govtypes.ModuleName)),
 		mint.NewAppModule(appCodec, app.MintKeeper, app.AccountKeeper, nil, app.GetSubspace(minttypes.ModuleName)),
 		distr.NewAppModule(appCodec, app.DistrKeeper, app.AccountKeeper, app.BankKeeper, app.StakingKeeper, app.GetSubspace(distrtypes.ModuleName)),
 		staking.NewAppModule(appCodec, app.StakingKeeper, app.AccountKeeper, app.BankKeeper, app.GetSubspace(stakingtypes.ModuleName)),
-
 		params.NewAppModule(app.ParamsKeeper),
 	)
 
@@ -455,7 +440,6 @@ func New(
 		genutiltypes.ModuleName,
 		paramstypes.ModuleName,
 		vestingtypes.ModuleName,
-		consensusparamtypes.ModuleName,
 	)
 
 	app.configurator = module.NewConfigurator(app.appCodec, app.MsgServiceRouter(), app.GRPCQueryRouter())

